cluster-autoscaler/processors/nodegroupset/balancing_processor.go

```diff
@@ -17,30 +17,54 @@ limitations under the License.
 package nodegroupset
 
 import (
-	"fmt"
 	"sort"
 
 	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
 	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	schedulercache "k8s.io/kubernetes/pkg/scheduler/cache"
 
 	"github.com/golang/glog"
 )
 
-// ScaleUpInfo contains information about planned scale-up of a single NodeGroup
-type ScaleUpInfo struct {
-	// Group is the group to be scaled-up
-	Group cloudprovider.NodeGroup
-	// CurrentSize is the current size of the Group
-	CurrentSize int
-	// NewSize is the size the Group will be scaled-up to
-	NewSize int
-	// MaxSize is the maximum allowed size of the Group
-	MaxSize int
+// BalancingNodeGroupSetProcessor tries to keep similar node groups balanced on scale-up.
+type BalancingNodeGroupSetProcessor struct {
+	Comparator NodeInfoComparator
 }
 
-// String is used for printing ScaleUpInfo for logging, etc
-func (s ScaleUpInfo) String() string {
-	return fmt.Sprintf("{%v %v->%v (max: %v)}", s.Group.Id(), s.CurrentSize, s.NewSize, s.MaxSize)
+// FindSimilarNodeGroups returns a list of NodeGroups similar to the given one.
+// Two groups are similar if the NodeInfos for them compare equal using IsNodeInfoSimilar.
+func (b *BalancingNodeGroupSetProcessor) FindSimilarNodeGroups(context *context.AutoscalingContext, nodeGroup cloudprovider.NodeGroup,
+	nodeInfosForGroups map[string]*schedulercache.NodeInfo) ([]cloudprovider.NodeGroup, errors.AutoscalerError) {
+
+	result := []cloudprovider.NodeGroup{}
+	nodeGroupId := nodeGroup.Id()
+	nodeInfo, found := nodeInfosForGroups[nodeGroupId]
+	if !found {
+		return []cloudprovider.NodeGroup{}, errors.NewAutoscalerError(
+			errors.InternalError,
+			"failed to find template node for node group %s",
+			nodeGroupId)
+	}
+	for _, ng := range context.CloudProvider.NodeGroups() {
+		ngId := ng.Id()
+		if ngId == nodeGroupId {
+			continue
+		}
+		ngNodeInfo, found := nodeInfosForGroups[ngId]
+		if !found {
+			glog.Warningf("Failed to find nodeInfo for group %v", ngId)
+			continue
+		}
+		comparator := b.Comparator
+		if comparator == nil {
```

Comment:
I would rather not have this logic here and treat Comparator as obligatory paremetrization.